interp/src/stk_env.rs

```diff
@@ -0,0 +1,809 @@
+    /// ```
+    pub fn set(&mut self, k: K, v: V) {
+        self.head.insert(k, v);
+    }
+
+    /// ```text
+    /// Returns a new Smoosher and mutates [self]. The new Smoosher has a new scope
+    /// as [head] and all of (pre-mutation) [self] as [tail]. [Self] has a fresh scope pushed onto
+    /// it. Invariant this method enforces: you cannot mutate a scope that has children
+    /// forks (you can only mutate the fresh scope applied atop it)
+    /// ```
+    /// # Examples
+    /// ## Pictorial Example
+    /// ```text
+    /// [A]
+    /// ```
+    /// let B = A.fork();
+    /// ```text
+    /// [B] [A'] <- Both B and A' point to empty nodes spawned off A
+    ///  |   |
+    ///   \ /
+    ///    |
+    ///   [A]
+    /// ```
+    /// ## Code Example
+    /// ```rust
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hi!", 1);
+    /// let mut b = a.fork();
+    /// a.set("hey", 2);
+    /// b.set("hey", 3);
+    /// assert_eq!(*a.get(&"hi!").unwrap(), *b.get(&"hi!").unwrap());
+    /// assert_eq!(*a.get(&"hey").unwrap(), 2);
+    /// assert_eq!(*b.get(&"hey").unwrap(), 3);
+    /// ```
+    pub fn fork(&mut self) -> Self {
+        //first save self's head, and replace it with a clean HM
+        let old_head = mem::replace(&mut self.head, HashMap::new()); //can replace with mem::take()
+                                                                     //create tail to both Self and the new Smoosher
+        let new_tail = self.tail.push(old_head);
+        //update Self and create new Smoosher
+        self.tail = new_tail.clone(); //will this die after the end of [fork]? no
+        Smoosher {
+            head: HashMap::new(),
+            tail: new_tail,
+        }
+    }
+
+    ///```text
+    ///Pushes a new, empty scope onto [self]. Doing so has no effect on the
+    ///bindings in [self], until a new [set] is called
+    ///```
+    /// # Example
+    /// ```rust
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hi!", 1);
+    /// a.new_scope();
+    /// a.set("hi!", 2);
+    /// assert_eq!(*a.get(&"hi!").unwrap(), 2);
+    /// ```
+    pub fn new_scope(&mut self) {
+        let old_head = mem::replace(&mut self.head, HashMap::new());
+        self.tail = self.tail.push(old_head);
+    }
+
+    /// ```text
+    /// Returns a pointer to the newest scope of this Smoosher. A Smoosher upon
+    /// instantiation will always have 1 empty HM as its first scope, so this
+    /// method will always meaningfully return.
+    /// ```
+    /// # Example
+    /// ```
+    /// use interp::stk_env::Smoosher;
+    /// let mut a = Smoosher::new();
+    /// a.set("hi!", 1);
+    /// a.set("bye", 0);
+    /// let hm = a.top();
+    /// assert_eq!(hm.len(), 2);
+    /// assert_eq!(*hm.get(&"hi!").unwrap(), 1);
+    /// assert_eq!(*hm.get(&"bye").unwrap(), 0);
+    /// ```
+    pub fn top(&self) -> &HashMap<K, V> {
+        &self.head
+    }
```

Comment:
I'm not entirely convinced that we need this method, but there's no harm in having it for now. Mostly just making this comment to remind myself to think on this some more

Reply:
okay, let us know. Or feel free to delete it yourself ofc

Reply:
I don't think there's any harm in keeping it around for the moment